src/lightning_fabric/strategies/deepspeed.py

```diff
@@ -365,24 +366,123 @@ def module_sharded_context(self) -> Generator[None, None, None]:
     def save_checkpoint(
         self, path: _PATH, state: Dict[str, Union[Module, Optimizer, Any]], storage_options: Optional[Any] = None
     ) -> None:
-        raise NotImplementedError
+        """Save model, optimizer, and other state in a checkpoint directory.
+
+        Args:
+            path: A path to where the files should be saved
+            state: A dictionary with contents to be saved. If the dict contains modules or optimizers, their
+                state-dict will be retrieved and converted automatically.
+            storage_options: Unused by this strategy, since it doesn't use a ``CheckpointIO`` plugin.
+
+        Raises:
+            TypeError:
+                If the unused ``storage_options`` gets passed.
+            ValueError:
+                When no :class:`deepspeed.DeepSpeedEngine` objects were found in the state, or when multiple
+                :class:`deepspeed.DeepSpeedEngine` objects were found.
+        """
+        if storage_options is not None:
+            raise TypeError(
+                "`DeepSpeedStrategy.save_checkpoint(..., storage_options=...)` is not supported because"
+                " `DeepSpeedStrategy` does not use the `CheckpointIO`."
+            )
+
+        engines = _get_deepspeed_engines_from_state(state)
```

Comment:
I think this is misleading. The engine is used to save but it doesn't save itself, only it's components: https://github.com/microsoft/DeepSpeed/blob/867da307d047c8122454ac8b0880cd23d38949e7/deepspeed/runtime/engine.py#L3276

So asking to pass the engine through the `state` makes it look like we are going to save it, when in reality it saves what's inside of it. Why not use the engine reference we have in `self` instead?

Reply:
Answer: My suggestion wouldn't work in examples that use multiple wrapped modules.